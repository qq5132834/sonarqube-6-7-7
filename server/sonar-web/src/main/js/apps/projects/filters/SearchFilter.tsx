/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { translate, translateWithParameters } from '../../../helpers/l10n';

interface Props {
  className?: string;
  handleSearch: (userString?: string) => void;
  query: { search?: string | undefined };
}

interface State {
  userQuery?: string;
}

export default class SearchFilter extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { userQuery: props.query.search };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (
      this.props.query.search === this.state.userQuery &&
      nextProps.query.search !== this.props.query.search
    ) {
      this.setState({ userQuery: nextProps.query.search || '' });
    }
  }

  handleQueryChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { value } = event.currentTarget;
    this.setState({ userQuery: value });
    if (!value || value.length >= 2) {
      this.props.handleSearch(value);
    }
  };

  render() {
    const { userQuery } = this.state;
    const shortQuery = userQuery != null && userQuery.length === 1;
    return (
      <div className={this.props.className}>
        <input
          type="search"
          value={userQuery || ''}
          placeholder={translate('projects.search')}
          onChange={this.handleQueryChange}
          autoComplete="off"
        />
        {shortQuery && (
          <span className="note spacer-left">{translateWithParameters('select2.tooShort', 2)}</span>
        )}
      </div>
    );
  }
}
