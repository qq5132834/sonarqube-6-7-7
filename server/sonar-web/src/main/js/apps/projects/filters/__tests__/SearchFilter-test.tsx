/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import SearchFilter from '../SearchFilter';
import { change } from '../../../../helpers/testUtils';

it('should render correctly without any search query', () => {
  const wrapper = shallow(<SearchFilter handleSearch={jest.fn()} query={{}} />);
  expect(wrapper).toMatchSnapshot();
});

it('should render with a search query', () => {
  const wrapper = shallow(<SearchFilter handleSearch={jest.fn()} query={{ search: 'foo' }} />);
  expect(wrapper).toMatchSnapshot();
});

it('should display a help message when there is less than 2 characters', () => {
  const wrapper = shallow(<SearchFilter handleSearch={jest.fn()} query={{ search: 'a' }} />);
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ userQuery: 'foo' });
  expect(wrapper).toMatchSnapshot();
});

it('searches', () => {
  const handleSearch = jest.fn();
  const wrapper = shallow(<SearchFilter handleSearch={handleSearch} query={{}} />);

  change(wrapper.find('input'), 'a');
  expect(handleSearch).not.toBeCalled();

  change(wrapper.find('input'), 'abc');
  expect(handleSearch).toBeCalledWith('abc');
});

it('updates state to new props', () => {
  const wrapper = shallow(<SearchFilter handleSearch={jest.fn()} query={{ search: 'abc' }} />);
  expect(wrapper.state()).toEqual({ userQuery: 'abc' });
  wrapper.setProps({ query: { search: 'def' } });
  expect(wrapper.state()).toEqual({ userQuery: 'def' });
});
