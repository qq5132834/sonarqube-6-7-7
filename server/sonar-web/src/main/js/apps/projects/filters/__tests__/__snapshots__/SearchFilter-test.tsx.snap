// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a help message when there is less than 2 characters 1`] = `
<div>
  <input
    autoComplete="off"
    onChange={[Function]}
    placeholder="projects.search"
    type="search"
    value="a"
  />
  <span
    className="note spacer-left"
  >
    select2.tooShort.2
  </span>
</div>
`;

exports[`should display a help message when there is less than 2 characters 2`] = `
<div>
  <input
    autoComplete="off"
    onChange={[Function]}
    placeholder="projects.search"
    type="search"
    value="foo"
  />
</div>
`;

exports[`should render correctly without any search query 1`] = `
<div>
  <input
    autoComplete="off"
    onChange={[Function]}
    placeholder="projects.search"
    type="search"
    value=""
  />
</div>
`;

exports[`should render with a search query 1`] = `
<div>
  <input
    autoComplete="off"
    onChange={[Function]}
    placeholder="projects.search"
    type="search"
    value="foo"
  />
</div>
`;
